src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -545,137 +571,319 @@ private RegexNode ReduceSet()
-        /// intervening operators can be flattened:
-        ///
-        /// a|b|c|def|g|h -> [a-c]|def|[gh]
-        /// apple|(?:orange|pear)|grape -> apple|orange|pear|grape
-        /// </summary>
+        /// <summary>Optimize an alternation.</summary>
         private RegexNode ReduceAlternation()
         {
-            int childCount = ChildCount();
-            if (childCount == 0)
+            switch (ChildCount())
             {
-                return new RegexNode(Nothing, Options);
-            }
+                case 0:
+                    return new RegexNode(Nothing, Options);
 
-            if (childCount == 1)
-            {
-                return Child(0);
-            }
+                case 1:
+                    return Child(0);
 
-            bool wasLastSet = false;
-            bool lastNodeCannotMerge = false;
-            RegexOptions optionsLast = 0;
-            RegexOptions optionsAt;
-            int i;
-            int j;
-            RegexNode at;
-            RegexNode prev;
+                default:
+                    ReduceSingleLetterAndNestedAlternations();
+                    RegexNode newThis = StripEnation(Nothing);
+                    return newThis != this ? newThis : ExtractCommonPrefix();
+            }
 
-            List<RegexNode> children = (List<RegexNode>)Children!;
-            for (i = 0, j = 0; i < children.Count; i++, j++)
+            // This function performs two optimizations:
+            // - Single-letter alternations can be replaced by faster set specifications
+            //   e.g. "a|b|c|def|g|h" -> "[a-c]|def|[gh]"
+            // - Nested alternations with no intervening operators can be flattened:
+            //   e.g. "apple|(?:orange|pear)|grape" -> "apple|orange|pear|grape"
+            void ReduceSingleLetterAndNestedAlternations()
             {
-                at = children[i];
+                bool wasLastSet = false;
+                bool lastNodeCannotMerge = false;
+                RegexOptions optionsLast = 0;
+                RegexOptions optionsAt;
+                int i;
+                int j;
+                RegexNode at;
+                RegexNode prev;
+
+                List<RegexNode> children = (List<RegexNode>)Children!;
+                for (i = 0, j = 0; i < children.Count; i++, j++)
+                {
+                    at = children[i];
 
-                if (j < i)
-                    children[j] = at;
+                    if (j < i)
+                        children[j] = at;
 
-                while (true)
-                {
-                    if (at.Type == Alternate)
+                    while (true)
                     {
-                        if (at.Children is List<RegexNode> atChildren)
+                        if (at.Type == Alternate)
                         {
-                            for (int k = 0; k < atChildren.Count; k++)
+                            if (at.Children is List<RegexNode> atChildren)
```

Comment:
Consider more comments throughout this method? Maybe with examples?

Reply:
This code is all pre-existing; only the indentation changed.  We can go through and figure out what it's doing and add comments separately.